### PR TITLE
travel: added option to search for next line

### PR DESCRIPTION
**Pull request by Adman**
* also did a little janitoring to prevent
  same function calls on multiple places

Signed-off-by: Adrian Matejov <a.matejov@centrum.sk>

### CODE REVIEW

**Comment by Adman** [comment]
@pepol @mrshu

File: travel.py [hunk 1]
```diff
@@ -9,6 +9,44 @@
...
+    """Sends output"""
+    out = 'Nothing found'
+
+    if date == '':
+        date = datetime.datetime.today().strftime('%d.%m.%Y')
+
+    if len(r):
+        out = r[0].__repr__()
+        try:
+            time = r[0].lines[0].departure
```
Comment (mrshu) [inline travel.py]: @Adman this seriously smells like Java.

File: travel.py [hunk 2]
```diff
@@ -9,6 +9,44 @@
 sys.setdefaultencoding("utf-8")
 
 
+# 'nick' : ['dep', 'dest', 'time', 'date']
+searched = {}
+
+
+def send(nick, dep, dest, date, r=None):
```
Comment (mrshu) [inline travel.py]: @Adman could we use a bit more descriptive name perhaps?

File: travel.py [hunk 3]
```diff
@@ -9,6 +9,44 @@
 sys.setdefaultencoding("utf-8")
 
 
+# 'nick' : ['dep', 'dest', 'time', 'date']
```
Comment (mrshu) [inline travel.py]: @Adman this inherently assumes that you are always looking for the next means of transportation of the last searched result. That is probably wrong as I can certainly think of situations where you'd search for someting in one channel and want to type `!bus next` in another one (in a different discussion).

I believe it is safe to ignore this for now but we should keep it in mind in general.

File: travel.py (hunk 2 above)
Comment (Adman) [inline travel.py]: @mrshu name for function or for parameters?

Comment (mrshu) [inline travel.py]: @Adman the name of the function is extremely general. `send` can really mean anything.

**Comment by Adman** [comment]
@mrshu Name of the function is now more understandable

File: travel.py [hunk 4]
```diff
@@ -9,6 +9,44 @@
...
+searched = {}
+
+
+def send_output(nick, dep, dest, date, r=None):
```
Comment (mrshu) [inline travel.py]: @Adman what is `r`? Should it be something like `result`?

**Comment by Adman** [comment]
@mrshu There you go

File: travel.py [hunk 5]
```diff
@@ -9,6 +9,44 @@
...
+    out = 'Nothing found'
+
+    if date == '':
+        date = datetime.datetime.today().strftime('%d.%m.%Y')
+
+    if len(result):
+        out = result[0].__repr__()
+        try:
+            time = result[0].lines[0].departure
+        except AttributeError:
```
Comment (mrshu) [inline travel.py]: @Adman this still smells like Java. Can't you just check if the first element of result has a `lines` attribute?

File: travel.py [hunk 6]
```diff
@@ -9,6 +9,44 @@
...
+            time = result[0].drives[0].begin_time
```
Comment (mrshu) [inline travel.py]: @Adman the 0th element of drives is not guaranteed to exist.

File: travel.py [hunk 7]
```diff
@@ -9,6 +9,44 @@
...
+        searched[nick] = [dep, dest, time, date]
+
+    out = unicode(out.strip(codecs.BOM_UTF8), 'utf-8')
+    return out.encode('utf-8')
+
+
+def searched_incrementer(nick):
+    """Adds 1 minute to previous search departure"""
+    user = searched[nick]
```
Comment (mrshu) [inline travel.py]: @Adman this really makes no sense. If you are using `user[0]`, `user[1]` and `user[2]` why don't you just do 

```python
dept, dest, time, date = searched[nick]
```

and use the respective variables? This is as unreadable as it gets.

File: travel.py [hunk 8]
```diff
@@ -9,6 +9,44 @@
 sys.setdefaultencoding("utf-8")
 
 
+# 'nick' : ['dep', 'dest', 'time', 'date']
+searched = {}
```
Comment (mrshu) [inline travel.py]: @Adman should this `searched` thing be preserved between bot's restarts? If so, maybe using the storage might be a good idea.

File: travel.py [hunk 9]
```diff
@@ -9,6 +9,44 @@
...
+    """Sends output"""
+    out = 'Nothing found'
+
+    if date == '':
+        date = datetime.datetime.today().strftime('%d.%m.%Y')
+
+    if len(result):
+        out = result[0].__repr__()
+        try:
+            time = result[0].lines[0].departure
```
Comment (mrshu) [inline travel.py]: @Adman the 0th element of lines is not guaranteed to exist.

File: travel.py [hunk 10]
```diff
@@ -9,6 +9,44 @@
...
+
+
+def send_output(nick, dep, dest, date, result=None):
+    """Sends output"""
+    out = 'Nothing found'
+
+    if date == '':
+        date = datetime.datetime.today().strftime('%d.%m.%Y')
+
+    if len(result):
```
Comment (mrshu) [inline travel.py]: @Adman this is abuse of notation. Please make this a boolean expression, something like `if len(result) >= 1`

File: travel.py [hunk 11]
```diff
@@ -29,8 +67,13 @@ def mhd(event):
 
     args = event.args
 
+    if len(args) and args[0] == 'next':
```
Comment (mrshu) [inline travel.py]: @Adman again, `len(args)` might work and look good in Perl, but this is Python.

File: travel.py [hunk 12]
```diff
@@ -105,18 +146,19 @@ def line_match(event, f, t, time=None, date=None, *args):
...
-    out = r[0].__repr__() if len(r) else "Nothing found"
-    out = unicode(out.strip(codecs.BOM_UTF8), 'utf-8')
-    return out.encode('utf-8')
+    return send_output(event.meta['nick'], f, t, date, result=r)
 
 
 def get_line(event, vehicle):
     """Searches for bus/train based on given vehicle argument"""
     args = event.args
+    if len(args) and args[0] == 'next':
```
Comment (mrshu) [inline travel.py]: @Adman once again `len(args)`

**Comment by mrshu** [comment]
@Adman a couple of nitpicks but generally it looks pretty good!

File: travel.py [hunk 13]
```diff
@@ -105,18 +146,19 @@ def line_match(event, f, t, time=None, date=None, *args):
...
+        if not event.meta['nick'] in searched:
+            return 'No next line'
+        args = searched_incrementer(event.meta['nick'])
 
     if len(args) < 2:
-        return
-
+        return 'Not enough arguments specified'
```
Comment (mrshu) [inline travel.py]: @Adman this is where the usage would be a nice thing to have.

File: travel.py (hunk 9 above)
Comment (Adman) [inline travel.py]: @mrshu In 99.9% cases it exists

Comment (mrshu) [inline travel.py]: @Adman so this will just blow up the whole script if it does not, right? Are you satisfied with that?

Comment (Adman) [inline travel.py]: @mrshu the It won't break the script, that's why I put here exception block

File: travel.py (hunk 8 above)
Comment (Adman) [inline travel.py]: @mrshu No, I think storing data like these ones is not neccessity in this plugin.

**Comment by Adman** [comment]
@mrshu all the nitpicks should be fixed

**Comment by mrshu** [comment]
@Adman looks fairly good, let's see what it does in prod.

**Pull request merged**